disnake/components.py

```diff
@@ -783,28 +839,401 @@ def to_dict(self) -> TextInputPayload:
+
+    def to_dict(self) -> SeparatorComponentPayload:
+        return {
+            "type": self.type.value,
+            "divider": self.divider,
+            "spacing": self.spacing.value,
+        }
+
+
+class Container(Component):
+    """Represents a container from the Discord Bot UI Kit (v2).
+
+    This is visually similar to :class:`Embed`\\s, and contains other components.
+
+    .. note::
+        The user constructible and usable type to create a
+        container is :class:`disnake.ui.Container`.
+
+    .. versionadded:: 2.11
+
+    Attributes
+    ----------
+    spoiler: :class:`bool`
+        Whether the container is marked as a spoiler. Defaults to ``False``.
+    components: List[Union[:class:`ActionRow`, :class:`Section`, :class:`TextDisplay`, :class:`MediaGallery`, :class:`FileComponent`, :class:`Separator`]]
+        The components in this container.
+    """
+
+    __slots__: Tuple[str, ...] = (
+        "_accent_colour",
+        "spoiler",
+        "components",
+    )
+
+    __repr_info__: ClassVar[Tuple[str, ...]] = (
+        "accent_colour",
+        "spoiler",
+        "components",
+    )
+
+    def __init__(self, data: ContainerComponentPayload) -> None:
+        self.type: Literal[ComponentType.container] = ComponentType.container
+        self._accent_colour: Optional[int] = data.get("accent_color")
+        self.spoiler: bool = data.get("spoiler", False)
+
+        components = [_component_factory(d) for d in data.get("components", [])]
+        self.components: List[ContainerChildComponent] = components  # type: ignore
+
+    def to_dict(self) -> ContainerComponentPayload:
+        payload: ContainerComponentPayload = {
+            "type": self.type.value,
+            "spoiler": self.spoiler,
+            "components": [child.to_dict() for child in self.components],
+        }
+
+        if self._accent_colour is not None:
+            payload["accent_color"] = self._accent_colour
+
+        return payload
+
+    @property
+    def accent_colour(self) -> Optional[Colour]:
+        """Optional[:class:`Colour`]: Returns the accent colour of the container.
+        An alias exists under ``accent_color``.
+        """
+        return Colour(self._accent_colour) if self._accent_colour is not None else None
+
+    @property
+    def accent_color(self) -> Optional[Colour]:
+        """Optional[:class:`Colour`]: Returns the accent color of the container.
+        An alias exists under ``accent_colour``.
+        """
+        return self.accent_colour
+
+
+# see ActionRowMessageComponent
+VALID_ACTION_ROW_MESSAGE_TYPES: Final = (
+    Button,
+    StringSelectMenu,
+    UserSelectMenu,
+    RoleSelectMenu,
+    MentionableSelectMenu,
+    ChannelSelectMenu,
+)
+
+
+def _walk_internal(component: Component, seen: Set[Component]) -> Iterator[Component]:
+    if component in seen:
+        # prevent infinite recursion if anyone manages to nest a component in itself
```

Comment:
What it actually does is skipping a component that has already been referenced.
That's not exactly the same as recursion, as a component could be reused between different containers/rows
It *is* an error if the component has a custom_id/id set, but for non-interactive components with omitted id it's technically fine